src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,580 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Result;
+use crate::token::{Token, TokenProvider};
+use tokio::sync::watch;
+use tokio::time::{Duration, Instant, sleep, sleep_until};
+
+const NORMAL_REFRESH_SLACK: Duration = Duration::from_secs(240);
+const SHORT_REFRESH_SLACK: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Clone)]
+pub(crate) struct TokenCache {
+    rx_token: watch::Receiver<Option<Result<Token>>>,
+}
+
+// TODO(#1552): Use the token cache in all creds
+#[allow(dead_code)]
+impl TokenCache {
+    pub(crate) fn new<T>(inner: T) -> Self
+    where
+        T: TokenProvider + Send + Sync + 'static,
+    {
+        let (tx_token, rx_token) = watch::channel::<Option<Result<Token>>>(None);
+
+        tokio::spawn(async move {
+            refresh_task(inner, tx_token).await;
+        });
+
+        Self { rx_token }
+    }
+}
+
+#[async_trait::async_trait]
+impl TokenProvider for TokenCache {
+    async fn get_token(&self) -> Result<Token> {
+        let mut rx = self.rx_token.clone();
+        let token_result = rx.borrow_and_update().clone();
+
+        if let Some(token_result) = token_result {
+            match token_result {
+                Ok(token) => match token.expires_at {
+                    None => return Ok(token),
+                    Some(e) => {
+                        if e < Instant::now().into_std() {
+                            // Expired token, wait for refresh
+                            return wait_for_next_token(rx).await;
+                        } else {
+                            // valid token
+                            return Ok(token);
+                        }
+                    }
+                },
+                // An error in the result is still a valid result to propagate to the client library
+                Err(e) => Err(e),
+            }
+        } else {
+            wait_for_next_token(rx).await
+        }
+    }
+}
+
+async fn wait_for_next_token(
+    mut rx_token: watch::Receiver<Option<Result<Token>>>,
+) -> Result<Token> {
+    rx_token.changed().await.unwrap();
+    let token_result = rx_token.borrow().clone();
+
+    token_result.expect("There should always be a token or error in the channel after changed()")
+}
+
+async fn refresh_task<T>(token_provider: T, tx_token: watch::Sender<Option<Result<Token>>>)
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    loop {
+        let token_result = token_provider.get_token().await;
+
+        let _ = tx_token.send(Some(token_result.clone()));
+
+        match token_result {
+            Ok(new_token) => {
+                if let Some(expiry) = new_token.expires_at {
+                    let time_until_expiry = expiry.duration_since(Instant::now().into_std());
```

Comment:
I should have noticed this earlier. Here you  need: https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_duration_since